File: src/strategies/perps.js
// nexus/src/strategies/perps.js — Perpetuals vs Spot Arbitrage Strategy

const MIN_SAFETY_FACTOR = 0.35; // net/gross must be ≥ 35%

/**
 * Finds arbitrage between spot sources and the perpetual (funding rate spread).
 * The perp price diverges from spot during high funding — we exploit that gap.
 *
 * @param {string} symbol
 * @param {Array}  spotSources — array of spot { price, exchange, fee }
 * @param {object|null} perpSource — { price, exchange, fee } from MEXC perps
 * @param {number} maxSpreadPct — volatility guard
 * @param {number} topN — return top N opportunities (default 1, legacy behavior)
 * @returns {Array|object|null} Array of top opportunities or single best (backward compatible)
 */
export function scanPerps(symbol, spotSources, perpSource, maxSpreadPct, topN = 1) {
  if (!perpSource || spotSources.length < 1) return topN === 1 ? null : [];

  const allSources = [...spotSources, perpSource];
  const prices     = allSources.map(s => s.price);
  const priceMin   = Math.min(...prices);
  const priceMax   = Math.max(...prices);
  const spread     = ((priceMax - priceMin) / priceMin) * 100;
  if (spread > maxSpreadPct) return topN === 1 ? null : [];

  const opportunities = [];
  for (const spot of spotSources) {
    for (const [buy, sell] of [[spot, perpSource], [perpSource, spot]]) {
      if (sell.price <= buy.price) continue;

      const grossPct    = ((sell.price - buy.price) / buy.price) * 100;
      const totalFeePct = (buy.fee + sell.fee) * 100;
      const netPct      = grossPct - totalFeePct;
      if (netPct <= 0) continue;

      const safetyFactor = netPct / grossPct;
      if (safetyFactor < MIN_SAFETY_FACTOR) continue;

      opportunities.push({
        strategy:     'perps',
        symbol,
        buyExchange:  buy.exchange,
        sellExchange: sell.exchange,
        buyPrice:     buy.price,
        sellPrice:    sell.price,
        grossPct,
        netPct,
        safetyFactor,
        direction:    `${buy.exchange.toUpperCase()}→${sell.exchange.toUpperCase()}`,
        isPerp:       true,
        timestamp:    Date.now(),
      });
    }
  }

  // Sort by net profit descending
  opportunities.sort((a, b) => (b.netPct ?? 0) - (a.netPct ?? 0));

  // Return top N (or single best for backward compatibility)
  const top = opportunities.slice(0, topN);
  return topN === 1 ? (top.length > 0 ? top[0] : null) : top;
}

File: agents/internal/strategies/perps/perps.go
// Package perps implements the perpetuals-vs-spot arbitrage strategy.
// When a perpetual's price diverges from spot (e.g. due to high funding),
// a delta-neutral position (long spot + short perp) captures the spread.
package perps

import (
	"strings"

	"github.com/zedanazad43/UltimateArbitrageHFT/hft/internal/feeds"
	"github.com/zedanazad43/UltimateArbitrageHFT/hft/internal/strategies/cex"
)

const minSafetyFactor = 0.4

// Scan evaluates all spot sources against a single perp source and returns
// the most profitable direction (spot→perp or perp→spot).
func Scan(symbol string, spotSources []feeds.PriceSource, perpSource *feeds.PerpData, maxSpreadPct float64) *cex.Opportunity {
	if perpSource == nil || len(spotSources) == 0 {
		return nil
	}

	// Spread guard across all sources (spot + perp).
	priceMin, priceMax := perpSource.Price, perpSource.Price
	for _, s := range spotSources {
		if s.Price < priceMin {
			priceMin = s.Price
		}
		if s.Price > priceMax {
			priceMax = s.Price
		}
	}
	if priceMin <= 0 {
		return nil
	}
	spread := ((priceMax - priceMin) / priceMin) * 100
	if spread > maxSpreadPct {
		return nil
	}

	perpPS := feeds.PriceSource{
		Price:    perpSource.Price,
		Exchange: perpSource.Exchange,
		Fee:      perpSource.Fee,
	}

	var best *cex.Opportunity
	for _, spot := range spotSources {
		for _, pair := range [][2]feeds.PriceSource{{spot, perpPS}, {perpPS, spot}} {
			buy, sell := pair[0], pair[1]
			if sell.Price <= buy.Price {
				continue
			}
			grossPct := ((sell.Price - buy.Price) / buy.Price) * 100
			totalFeePct := (buy.Fee + sell.Fee) * 100
			netPct := grossPct - totalFeePct
			if netPct <= 0 {
				continue
			}
			safetyFactor := netPct / grossPct
			if safetyFactor < minSafetyFactor {
				continue
			}
			if best == nil || netPct > best.NetPct {
				best = &cex.Opportunity{
					Strategy:     "perps",
					Symbol:       symbol,
					BuyExchange:  buy.Exchange,
					SellExchange: sell.Exchange,
					BuyPrice:     buy.Price,
					SellPrice:    sell.Price,
					GrossPct:     grossPct,
					NetPct:       netPct,
					SafetyFactor: safetyFactor,
					Direction:    strings.ToUpper(buy.Exchange) + "→" + strings.ToUpper(sell.Exchange),
					IsPerp:       true,
				}
			}
		}
	}
	return best
}
